### PR TITLE
[FIX] stock: should_bypass_reservation for internal scrap locations

### PR DESCRIPTION
**Pull request by Kosaaaaa**
Description of the issue/feature this PR addresses:

Current behavior before PR: It is possible to skip reservation on internal scrap locations and have negative quantity on this location.

Desired behavior after PR is merged: Make reservation on scrap internal location to prevent this situation.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by MiquelRForgeFlow** [comment]
Read this [guidelines](https://www.odoo.com/documentation/15.0/contributing/development/git_guidelines.html).

**Comment by Kosaaaaa** [comment]
> Read this [guidelines](https://www.odoo.com/documentation/15.0/contributing/development/git_guidelines.html).

Done

**Comment by pfertyk** [comment]
Hi @Kosaaaaa and thank you for contributing! I can see that some tests are failing, could you please check if it's caused by your changes? And, since you've added a bit of logic to Odoo, could you also please add a test?

**Comment by Kosaaaaa** [comment]
Hi @pfertyk! Thank you for your support, I've just rebased with 15.0

**Comment by Kosaaaaa** [comment]
@pfertyk finally all is green 😁

**Comment by pfertyk** [comment]
Thanks! Let me double check the new logic with my colleagues :) Meanwhile, could you please squash your commits, so we end up with a single one? (you can use git fixup locally and then force push)

File: addons/stock/tests/test_robustness.py [hunk 1]
```diff
@@ -113,6 +114,53 @@ def test_location_usage(self):
...
+        # assert the reservation
+        self.assertEqual(quant.reserved_quantity, 1)  # reservation is not bypassed in scrap location
+        self.assertEqual(move1.product_qty, 1)
+
+        # change the stock usage
+        with self.assertRaises(UserError):
+            with self.cr.savepoint():
+                test_stock_location.scrap_location = False
+
+        # unreserve
```
Comment (pfertyk) [inline addons/stock/tests/test_robustness.py:161]: I am not sure if this part is required :thinking: Changes made in the database are reverted after each test, so unless there is a good reason to keep this line, I would remove it.

Comment (Kosaaaaa) [reply addons/stock/tests/test_robustness.py]: done

**Comment by Kosaaaaa** [comment]
> Thanks! Let me double check the new logic with my colleagues :) Meanwhile, could you please squash your commits, so we end up with a single one? (you can use git fixup locally and then force push)

done

**Comment by kharkot** [comment]
Hello, if/when fix will be approved. Our customer is waiting for it.

File: odooE.conf [hunk 2]
```diff
@@ -0,0 +1,12 @@
+[options]
```
Comment (pfertyk) [inline odooE.conf:1]: @Kosaaaaa Are you sure that you wanted to add this file to the codebase? ;)

**Comment by pfertyk** [comment]
Hi @kharkot ! I still need to confirm if this is the logic we want to become a part of Odoo codebase. From the info I've gathered so far it seems that we actually want to keep the current behavior, but I am double-checking this.

**Comment by kharkot** [comment]
Hi @pfertyk but current behavior have no logic, if there is no logic it is a bug - quite simple - don't you think so?

**Comment by pfertyk** [comment]
Hi @kharkot !

The current behavior is that scrap locations do not handle reservations. That's because they are considered like trashcan locations: if a product ends up there it means that it was removed (or recycled). That's why there should not be a situation like:
* a non-internal scrap location
* a reservation made from a scrap location

For more information, please check [this entry](https://www.odoo.com/documentation/master/applications/inventory_and_mrp/manufacturing/workflows/scrap_manufacturing.html) in our docs. There is also a quick explanation [here](https://www.odoo.com/documentation/17.0/applications/inventory_and_mrp/inventory/warehouses_storage/inventory_management/scrap_inventory.html).

In short: products put in scrap cease to exist. If there is a chance that disassembling them would create some useful spare parts, then  "unbuild" operation should be used instead of "scrap".

I've consulted this with two colleagues and we decided not to merge this PR.

Having said that, I really appreciate your work @Kosaaaaa ! You reacted quickly, made all the required fixes, and added a test, which is always very welcome! :heart: I am sorry that your PR will not be included in Odoo, but I hope that it will not discourage you and I'll get a chance to review more of our code in the future :wink:

**Comment by kharkot** [comment]
@pfertyk The situation you described is OK but only when the scrap location is virtual (from documentation: "By default, scrapping a component or finished product removes it from physical inventory and places it in a virtual location titled Virtual Locations/Scrap. A virtual location is not a physical space, but rather a designation in Odoo that is used to track items that are no longer in physical inventory.")
In case scrap location is internal, products are  still  available to use and are presented as ON HAND quantities, forecasted quantities

So question for you and your colleagues - why there is  possible to define scrap location as internal location (not virtual) .

Let's make Odoo better!

**Comment by pfertyk** [comment]
Hi @kharkot ! Currently, Odoo offers the flexibility to mark an internal location as a scrap location, if there is any reason to do so. In that case, however, we do not make any adjustments in expected reservations to accommodate for that.

In most cases, a scrap location should be configured as virtual. This results in a more predictable behavior, and is also more intuitive, since scrapped products are considered as nonexistent. Despite other options being available, this configuration should be used in most cases.

Is it possible for you and your team to switch to a virtual scrap location? Would it resolve the issue?
